A docs update with production guide + restructured advanced guide (#2232)
zenml-io/zenml: docs/book/user-guide/production-guide/connect-code-repository.md

---
description: >-
  Connect a Git repository to ZenML to track code changes and collaborate on MLOps projects.
---

# Connect your code repository

> Throughout the lifecycle of a MLOps pipeline, it can get quite tiresome to always wait for a docker build everytime after running a pipeline (even if the local Docker cache is used). However, there is a way to just have one pipeline build and keep reusing it until a change to the pipeline environment is made: by connecting a code repository.

Reviewer: ```suggestion
Throughout the lifecycle of a MLOps pipeline, it can get quite tiresome to always wait for a docker build every time after running a pipeline (even if the local Docker cache is used). However, there is a way to just have one pipeline build and keep reusing it until a change to the pipeline environment is made: by connecting a code repository.
```

> Throughout the lifecycle of a MLOps pipeline, it can get quite tiresome to always wait for a docker build every time after running a pipeline (even if the local Docker cache is used). However, there is a way to just have one pipeline build and keep reusing it until a change to the pipeline environment is made: by connecting a code repository.

Reviewer 2: ```suggestion
Throughout the lifecycle of a MLOps pipeline, it can get quite tiresome to always wait for a docker build every time after running a pipeline (even if the local Docker cache is used). However, there is a way to just have one pipeline build and keep reusing it until a change to the pipeline environment is made: by connecting a code repository.
```
```suggestion
Throughout the lifecycle of a MLOps pipeline, it can get quite tiresome to always wait for a Docker build every time after running a pipeline (even if the local Docker cache is used). However, there is a way to just have one pipeline build and keep reusing it until a change to the pipeline environment is made: by connecting a code repository.
```

> With ZenML, connecting to a Git repository optimizes Docker build processes. It also has the added bonus of being a better way of managing repository changes and enabling better code collaboration. Here is how are flow changes when running a pipeline:

Reviewer: ```suggestion
With ZenML, connecting to a Git repository optimizes the Docker build processes. It also has the added bonus of being a better way of managing repository changes and enabling better code collaboration. Here is how the flow changes when running a pipeline:
```

<figure><img src="../../.gitbook/assets/run_with_repository.png" alt=""><figcaption><p>Sequence of events that happen when running a pipeline on a remote stack with a code repository</p></figcaption></figure>

1. You trigger a pipeline run on your local machine. ZenML parses the `@pipeline` function to determine the necessary steps.
2. The local client requests stack information from the ZenML server, which responds with the cloud stack configuration.

> 3. The local client detects that we're using a connected a code repository and requests the information from the git repo.

Reviewer: ```suggestion
3. The local client detects that we're using a code repository and requests the information from the git repo.
```
